File: src/test/java/com/wordonline/matching/server/service/GameServerManagementServiceTest.java
package com.wordonline.matching.server.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;

import com.wordonline.matching.server.client.GameServerClient;
import com.wordonline.matching.server.entity.Server;
import com.wordonline.matching.server.entity.ServerState;
import com.wordonline.matching.server.entity.ServerType;
import com.wordonline.matching.session.repository.ServerRepository;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

@ExtendWith(MockitoExtension.class)
class GameServerManagementServiceTest {

    private static final int FAILURE_THRESHOLD = 3;
    private static final String ALPHA_URL = "http://alpha:9090";
    private static final String BETA_URL = "http://beta:9090";

    @Mock
    private ServerRepository serverRepository;
    @Mock
    private GameServerClient gameServerClient;

    private ServerHealthRegistry serverHealthRegistry;
    private GameServerManagementService gameServerManagementService;

    @BeforeEach
    void setUp() {
        serverHealthRegistry = new ServerHealthRegistry(FAILURE_THRESHOLD);
        gameServerManagementService =
                new GameServerManagementService(serverRepository, gameServerClient, serverHealthRegistry);
    }

    private static Server server(long id, String domain, ServerState state) {
        Server server = new Server();
        ReflectionTestUtils.setField(server, "id", id);
        ReflectionTestUtils.setField(server, "protocol", "http");
        ReflectionTestUtils.setField(server, "domain", domain);
        ReflectionTestUtils.setField(server, "port", 9090);
        ReflectionTestUtils.setField(server, "state", state);
        ReflectionTestUtils.setField(server, "type", ServerType.GAME);
        return server;
    }

    private void stubDiscovery(Server... servers) {
        when(serverRepository.findAllByType(ServerType.GAME)).thenReturn(Flux.just(servers));
    }

    private void refresh() {
        StepVerifier.create(gameServerManagementService.refresh()).verifyComplete();
    }

    private List<Long> availableIds() {
        return gameServerManagementService.getAvailableServers().stream().map(Server::getId).toList();
    }

    @Test
    void 한_서버의_헬스체크_실패가_다른_서버의_상태_갱신을_막지_않는다() {
        stubDiscovery(
                server(1L, "alpha", ServerState.ACTIVE),
                server(2L, "beta", ServerState.ACTIVE)
        );
        when(gameServerClient.healthcheck(ALPHA_URL))
                .thenReturn(Mono.error(new IllegalStateException("boom")));
        when(gameServerClient.healthcheck(BETA_URL)).thenReturn(Mono.just(true));

        refresh();

        assertThat(availableIds()).containsExactly(2L);
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isEqualTo(1);
    }

    @Test
    void DB가_INACTIVE로_표시한_서버도_헬스체크에_성공하면_배정_대상이_된다() {
        stubDiscovery(server(1L, "alpha", ServerState.INACTIVE));
        when(gameServerClient.healthcheck(ALPHA_URL)).thenReturn(Mono.just(true));

        refresh();

        assertThat(availableIds()).containsExactly(1L);
    }

    @Test
    void DRAINING_서버는_헬스체크에_성공해도_배정_대상에서_제외된다() {
        stubDiscovery(
                server(1L, "alpha", ServerState.DRAINING),
                server(2L, "beta", ServerState.ACTIVE)
        );
        when(gameServerClient.healthcheck(ALPHA_URL)).thenReturn(Mono.just(true));
        when(gameServerClient.healthcheck(BETA_URL)).thenReturn(Mono.just(true));

        refresh();

        assertThat(availableIds()).containsExactly(2L);
        assertThat(gameServerManagementService.getAvailableServer())
                .map(Server::getId)
                .contains(2L);
    }

    @Test
    void 연속_실패가_임계값에_도달해야_사용_불가로_판정하고_성공_한_번에_즉시_복구한다() {
        stubDiscovery(server(1L, "alpha", ServerState.ACTIVE));

        when(gameServerClient.healthcheck(ALPHA_URL)).thenReturn(Mono.just(true));
        refresh();
        assertThat(availableIds()).containsExactly(1L);

        when(gameServerClient.healthcheck(ALPHA_URL)).thenReturn(Mono.just(false));
        refresh();
        assertThat(availableIds()).as("1회 실패로는 내리지 않는다").containsExactly(1L);
        refresh();
        assertThat(availableIds()).as("2회 실패로도 내리지 않는다").containsExactly(1L);
        refresh();
        assertThat(availableIds()).as("3회 연속 실패면 내린다").isEmpty();

        when(gameServerClient.healthcheck(ALPHA_URL)).thenReturn(Mono.just(true));
        refresh();
        assertThat(availableIds()).as("성공 1회면 즉시 복구한다").containsExactly(1L);
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isZero();
    }

    @Test
    void 한_번도_성공하지_못한_서버는_배정_대상이_아니다() {
        stubDiscovery(server(1L, "alpha", ServerState.ACTIVE));
        when(gameServerClient.healthcheck(ALPHA_URL)).thenReturn(Mono.just(false));

        refresh();

        assertThat(availableIds()).isEmpty();
    }

    @Test
    void 리로드해도_헬스_이력이_유지된다() {
        // every refresh re-reads the row and hands back a brand new Server instance,
        // so the failure counter can only survive if it is keyed by server id.
        when(serverRepository.findAllByType(ServerType.GAME))
                .thenAnswer(invocation -> Flux.just(server(1L, "alpha", ServerState.ACTIVE)));

        when(gameServerClient.healthcheck(ALPHA_URL)).thenReturn(Mono.just(true));
        refresh();

        when(gameServerClient.healthcheck(ALPHA_URL)).thenReturn(Mono.just(false));
        refresh();
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isEqualTo(1);
        refresh();
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isEqualTo(2);
        assertThat(availableIds()).containsExactly(1L);
        refresh();
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isEqualTo(FAILURE_THRESHOLD);
        assertThat(availableIds()).isEmpty();
    }

    @Test
    void 사라진_서버의_헬스_이력은_정리된다() {
        stubDiscovery(server(1L, "alpha", ServerState.ACTIVE));
        when(gameServerClient.healthcheck(ALPHA_URL)).thenReturn(Mono.just(true));
        refresh();
        assertThat(serverHealthRegistry.isHealthy(1L)).isTrue();

        serverHealthRegistry.replaceServers(List.of());

        assertThat(serverHealthRegistry.isHealthy(1L)).isFalse();
        assertThat(gameServerManagementService.getAvailableServers()).isEmpty();
    }
}

File: src/test/kotlin/com/wordonline/matching/server/service/GameServerManagementServiceTest.kt
package com.wordonline.matching.server.service

import com.wordonline.matching.server.client.GameServerClient
import com.wordonline.matching.server.config.GameServerProperties
import com.wordonline.matching.server.entity.Server
import com.wordonline.matching.server.entity.ServerState
import com.wordonline.matching.server.entity.ServerType
import com.wordonline.matching.session.repository.ServerRepository
import kotlinx.coroutines.test.runTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.doThrow
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.stub
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import reactor.core.publisher.Flux

class GameServerManagementServiceTest {

    private val failureThreshold = 3
    private val alphaUrl = "http://alpha:9090"
    private val betaUrl = "http://beta:9090"

    private val serverRepository: ServerRepository = mock()
    private val gameServerClient: GameServerClient = mock()

    private val properties = GameServerProperties(failureThreshold = failureThreshold)
    private val serverHealthRegistry = ServerHealthRegistry(properties)
    private val gameServerManagementService =
        GameServerManagementService(serverRepository, gameServerClient, serverHealthRegistry)

    private fun server(id: Long, domain: String, state: ServerState, internalBaseUrl: String? = null): Server =
        Server(
            id = id,
            protocol = "http",
            domain = domain,
            port = 9090,
            state = state,
            type = ServerType.GAME,
            internalBaseUrl = internalBaseUrl,
        )

    private fun stubDiscovery(vararg servers: Server) {
        // re-created per call: every refresh re-reads the row and hands back a brand new
        // Server instance, so health history can only survive if it is keyed by server id.
        whenever(serverRepository.findAllByType(ServerType.GAME))
            .thenAnswer { Flux.fromIterable(servers.toList()) }
    }

    private fun stubHealthcheck(serverUrl: String, healthy: Boolean) {
        gameServerClient.stub { onBlocking { healthcheck(serverUrl) } doReturn healthy }
    }

    private fun stubHealthcheckThrows(serverUrl: String, error: Throwable) {
        gameServerClient.stub { onBlocking { healthcheck(serverUrl) } doThrow error }
    }

    private fun availableIds(): List<Long> =
        gameServerManagementService.getAvailableServers().mapNotNull { it.id }

    @Test
    fun `한 서버의 헬스체크 실패가 다른 서버의 상태 갱신을 막지 않는다`() = runTest {
        stubDiscovery(
            server(1L, "alpha", ServerState.ACTIVE),
            server(2L, "beta", ServerState.ACTIVE),
        )
        stubHealthcheckThrows(alphaUrl, IllegalStateException("boom"))
        stubHealthcheck(betaUrl, true)

        gameServerManagementService.refresh()

        assertThat(availableIds()).containsExactly(2L)
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isEqualTo(1)
    }

    @Test
    fun `DB가 INACTIVE로 표시한 서버도 헬스체크에 성공하면 배정 대상이 된다`() = runTest {
        stubDiscovery(server(1L, "alpha", ServerState.INACTIVE))
        stubHealthcheck(alphaUrl, true)

        gameServerManagementService.refresh()

        assertThat(availableIds()).containsExactly(1L)
    }

    @Test
    fun `DRAINING 서버는 헬스체크에 성공해도 배정 대상에서 제외된다`() = runTest {
        stubDiscovery(
            server(1L, "alpha", ServerState.DRAINING),
            server(2L, "beta", ServerState.ACTIVE),
        )
        stubHealthcheck(alphaUrl, true)
        stubHealthcheck(betaUrl, true)

        gameServerManagementService.refresh()

        assertThat(availableIds()).containsExactly(2L)
        assertThat(gameServerManagementService.getAvailableServer()?.id).isEqualTo(2L)
    }

    @Test
    fun `연속 실패가 임계값에 도달해야 사용 불가로 판정하고 성공 한 번에 즉시 복구한다`() = runTest {
        stubDiscovery(server(1L, "alpha", ServerState.ACTIVE))

        stubHealthcheck(alphaUrl, true)
        gameServerManagementService.refresh()
        assertThat(availableIds()).containsExactly(1L)

        stubHealthcheck(alphaUrl, false)
        gameServerManagementService.refresh()
        assertThat(availableIds()).`as`("1회 실패로는 내리지 않는다").containsExactly(1L)
        gameServerManagementService.refresh()
        assertThat(availableIds()).`as`("2회 실패로도 내리지 않는다").containsExactly(1L)
        gameServerManagementService.refresh()
        assertThat(availableIds()).`as`("3회 연속 실패면 내린다").isEmpty()

        stubHealthcheck(alphaUrl, true)
        gameServerManagementService.refresh()
        assertThat(availableIds()).`as`("성공 1회면 즉시 복구한다").containsExactly(1L)
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isZero()
    }

    @Test
    fun `한 번도 성공하지 못한 서버는 배정 대상이 아니다`() = runTest {
        stubDiscovery(server(1L, "alpha", ServerState.ACTIVE))
        stubHealthcheck(alphaUrl, false)

        gameServerManagementService.refresh()

        assertThat(availableIds()).isEmpty()
    }

    @Test
    fun `리로드해도 헬스 이력이 유지된다`() = runTest {
        stubDiscovery(server(1L, "alpha", ServerState.ACTIVE))

        stubHealthcheck(alphaUrl, true)
        gameServerManagementService.refresh()

        stubHealthcheck(alphaUrl, false)
        gameServerManagementService.refresh()
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isEqualTo(1)
        gameServerManagementService.refresh()
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isEqualTo(2)
        assertThat(availableIds()).containsExactly(1L)
        gameServerManagementService.refresh()
        assertThat(serverHealthRegistry.consecutiveFailures(1L)).isEqualTo(failureThreshold)
        assertThat(availableIds()).isEmpty()
    }

    @Test
    fun `internal_base_url이 있는 서버는 healthcheck을 내부 주소로 보낸다`() = runTest {
        val internalUrl = "http://ac-game-alpha:8080"
        stubDiscovery(server(1L, "alpha", ServerState.ACTIVE, internalBaseUrl = internalUrl))
        stubHealthcheck(internalUrl, true)

        gameServerManagementService.refresh()

        assertThat(availableIds()).containsExactly(1L)
        verify(gameServerClient, never()).healthcheck(alphaUrl)
    }

    @Test
    fun `사라진 서버의 헬스 이력은 정리된다`() = runTest {
        stubDiscovery(server(1L, "alpha", ServerState.ACTIVE))
        stubHealthcheck(alphaUrl, true)
        gameServerManagementService.refresh()
        assertThat(serverHealthRegistry.isHealthy(1L)).isTrue()

        serverHealthRegistry.replaceServers(emptyList())

        assertThat(serverHealthRegistry.isHealthy(1L)).isFalse()
        assertThat(gameServerManagementService.getAvailableServers()).isEmpty()
    }
}
